### DISCUSSION TITLE
VPack as block storage

### DISCUSSION THREAD

**Discussion by dumblob** [erthink/libmdbx#340] [Q&A] [answered]
Has anyone thought about trying other low-level storage backends to experiment with efficiency & performance? Perhaps something like [VPack](https://github.com/arangodb/velocypack ) could be interesting to test.

I'm asking despite seeing libmdbx's internals being rather more tightly bound to the current format (which seems nearly identical to LMDB).

**Comment by erthink** [comment, marked as answer]
MDBX treat user's keys and data as opacity byte strings, except the [`MDBX_INTEGERKEY` and `MDBX_INTEGERDUP`](https://libmdbx.dqdkfa.ru/group__c__dbi.html#gafe3bddb297b3ab0d828a487c5726f76a) features. So nothing prevents you from using any serialization format, including noted VelocyPack.

On the other hand, using custom format for keys requires "custom comparators" which entails [certain restrictions](https://github.com/erthink/libmdbx/blob/master/mdbx.h#L3544-L3550). Therefore, it is [planned](https://github.com/erthink/libmdbx/issues/115) to implement minimal support for MessagePack.
However, I do think it is NOT reasonable to do same for VPack, since it is used rarely in comparison with MsgPack.

**Comment by dumblob** [comment]
Oh, pardon me for not being clear with my question.

I'm asking whether someone has tried to exchange the memory layout (and thus also the on-disk data layout) libmdbx uses **internally** for VPack or anything similar.

**Comment by erthink** [comment]
Suggest to close since no answer is expected.
